### PR TITLE
[IMP] stock: add generate button for lots

### PR DESCRIPTION
**Pull request by Sehnde**
This PR makes it possible to generate lot names/numbers in batch from operations in stock picking.

It uses the same `count` field as generate serials to ask for a specific quantity per lot.
A new field `totalReceived` is used to indicate how much is received in total. This is necessary
in cases of partial reception.

The value indicated in `count` will be assigned to as many lots as possible without exceeding
`totalReceived`, with the leftover assigned to a last lot.
The formula: `number_of_lots_generated = (totalReceived // quantity_per_lot) + 1` (if there's leftover).

The rest works the same as the generate serials flow.

task 3506686

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/static/src/widgets/lots_dialog.xml [hunk 1]
```diff
@@ -13,30 +13,59 @@
...
                         <div class="d-flex d-sm-contents">
                             <div class="o_cell flex-grow-0 flex-sm-grow-0 text-900 pe-3">
-                                <label class="o_form_label" for="next_serial_0">First SN</label>
+
+                                <t t-if="props.move.data.has_tracking === 'lot'">
+                                    <label class="o_form_label" for="next_serial_0">First Lot Number</label>
+                                </t>
+                                <t t-else="">
+                                    <label class="o_form_label" for="next_serial_0">First Serial Number</label>
+                                </t>
```
Comment (svs-odoo) [inline addons/stock/static/src/widgets/lots_dialog.xml:22]:
```suggestion
                                <label class="o_form_label" for="next_serial_0">
                                    <t t-if="props.move.data.has_tracking === 'lot'">First Lot Number</t>
                                    <t t-else="">First Serial Number</t>
                                </label>
```

File: addons/stock/static/src/widgets/lots_dialog.xml [hunk 2]
```diff
@@ -13,30 +13,59 @@
...
                         </div>
                         <div class="d-flex d-sm-contents">
                             <div class="o_cell flex-grow-0 flex-sm-grow-0 text-900 pe-3">
-                                <label class="o_form_label" for="next_serial_count_0">Number of SN</label>
+                                <t t-if="props.move.data.has_tracking === 'lot'">
+                                    <label class="o_form_label" for="next_serial_count_0">Quantity per Lot</label>
+                                </t>
+                                <t t-else="">
+                                    <label class="o_form_label" for="next_serial_count_0">Number of SN</label>
+                                </t>
```
Comment (svs-odoo) [inline addons/stock/static/src/widgets/lots_dialog.xml:37]: Same suggestion than previous comment

File: addons/stock/static/src/widgets/lots_dialog.xml [hunk 3]
```diff
@@ -13,30 +13,59 @@
...
+                                    <div name="next_serial_count" class="o_field_widget o_field_integer">
+                                        <input inputmode="numeric" t-ref="nextSerialCount" class="o_input" id="next_serial_count_0" type="text"/>
+                                    </div>
                                 </div>
+                                <span t-if="props.move.data.has_tracking === 'lot'" t-esc="props.move.data.product_uom[1]"/>
                             </div>
                         </div>
+                        <div class="d-flex d-sm-contents">
+                            <t t-if="props.move.data.has_tracking === 'lot'">
+                                <div class="o_cell flex-grow-0 flex-sm-grow-0 text-900 pe-3">
```
Comment (svs-odoo) [inline addons/stock/static/src/widgets/lots_dialog.xml:50]: You can set the `t-if` directly on the `div`
```suggestion
                        <div t-if="props.move.data.has_tracking === 'lot'" class="d-flex d-sm-contents">
                            <div class="o_cell flex-grow-0 flex-sm-grow-0 text-900 pe-3">
```

File: addons/stock/static/src/widgets/lots_dialog.xml [hunk 4]
```diff
@@ -47,7 +76,12 @@
                     <div class="grid o_inner_group">
                         <div class="d-flex d-sm-contents">
                             <div class="o_cell flex-grow-0 flex-sm-grow-0 text-900 pe-3">
-                                <label class="o_form_label" for="next_serial_0">Lots/Serial numbers</label>
+                                <t t-if="props.move.data.has_tracking==='lot'">
+                                    <label class="o_form_label" for="next_serial_0">Lot numbers</label>
+                                </t>
+                                <t t-else="">
+                                    <label class="o_form_label" for="next_serial_0">Serial numbers</label>
+                                </t>
```
Comment (svs-odoo) [inline addons/stock/static/src/widgets/lots_dialog.xml:84]: Same suggestion than previously

File: addons/stock/models/stock_move.py [hunk 5]
```diff
@@ -198,7 +198,7 @@ def _compute_display_assign_serial(self):
                 not move.origin_returned_move_id.id and
                 move.state not in ('done', 'cancel')
             )
-            move.display_assign_serial = move.has_tracking == 'serial' and move.display_import_lot
+            move.display_assign_serial = move.display_import_lot
```
Comment (svs-odoo) [inline addons/stock/models/stock_move.py:201]: Check if the override of `_compute_display_assign_serial` in `mrp_subcontracting` doesn't need to be rework (it skips move not tracked by serial, but I guess it should not skip move tracked by lot now)

**Review by svs-odoo — APPROVED** [review]
Some suggestions for cleaning/shorting the code but seems pretty good overall :)

File: addons/stock/models/stock_move.py [hunk 6]
```diff
@@ -881,6 +881,17 @@ def split_lots(self, lots):
     def action_generate_lot_line_vals(self, context, mode, first_lot, count, lot_text):
         assert mode in ('serial', 'import')
         default_vals = {}
+
+        def generate_lot_qty(quantity, qty_per_lot):
+            if qty_per_lot <= 0:
+                raise UserError(_("The quantity per lot should always be a positive value."))
+            line_count = int(quantity // qty_per_lot)
+            leftover = quantity % qty_per_lot
+            qty_array = [qty_per_lot for _ in range(line_count)]
```
Comment (svs-odoo) [inline addons/stock/models/stock_move.py:890]: There is always something cool to learn with python :nerd_face: 
```suggestion
            qty_array = [qty_per_lot] * line_count
```

File: addons/stock/models/stock_move.py [hunk 7]
```diff
@@ -891,13 +902,19 @@ def remove_prefix(text, prefix):
                 default_vals[remove_prefix(key, 'default_')] = context[key]
 
         vals_list = []
+        if default_vals['tracking'] == 'lot' and mode == 'serial':
+            lot_qties = generate_lot_qty(default_vals['quantity'], count)
+        else:
+            lot_qties = [1 for _ in range(count)]
```
Comment (svs-odoo) [inline addons/stock/models/stock_move.py:908]: Same here :)
```suggestion
            lot_qties = [1] * count
```

**Review by svs-odoo — APPROVED** [review]
I'm thinking... About the `type`, shouldn't we use `generate` and `import` instead of `serial` and `lot` ? It should be clearer.
Also, this parameter is named `type` in JS and `mode` in Python. It should be cool to have the same name in both side

**Comment by Sehnde** [comment]
@svs-odoo done

**Comment by svs-odoo** [comment]
robodoo r+

**Pull request closed without merging**